File: src/algorithms/Remove_Duplicates_from_Sorted_Array_II.py
# -*- coding: utf-8 -*-


class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def removeDuplicates(self, nums):
        if nums == []:
            return 0
        n_len, now = 0, 0
        tmp, pre = 0, nums[0]
        for num in nums:
            if pre == num:
                tmp += 1
            else:
                cnt = min(tmp, 2)
                n_len += cnt
                for j in range(cnt):
                    nums[now] = pre
                    now += 1
                tmp, pre = 1, num
        cnt = min(tmp, 2)
        n_len += cnt
        for j in range(cnt):
            nums[now] = nums[-1]
            now += 1
        return n_len


if __name__ == '__main__':
    solution = Solution()
    print solution.removeDuplicates([1, 1, 1, 2, 2, 3])
    print solution.removeDuplicates([1])
    print solution.removeDuplicates([1, 2])

File: src/algorithms/cpp/Remove_Duplicates_from_Sorted_Array_II.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int removeDuplicates(vector <int>& nums) {
        if (SZ(nums) == 0)  return 0;
        int pos = 0, pre = -1, cnt = 0;
        bool tag = false;
        while (pos < SZ(nums)) {
            if (!tag) {
                pre = nums[pos];
                cnt = 1;
                tag = true;
                ++ pos;
            }
            else {
                if (nums[pos] == pre) {
                    if (cnt == 2) {
                        nums.erase(nums.begin() + pos);
                    }
                    else {
                        ++ cnt;
                        ++ pos;
                    }
                }
                else {
                    pre = nums[pos];
                    cnt = 1;
                    ++ pos;
                }
            }
        }
        return SZ(nums);
    }
};

int main() {
    vector <int> input_v = {1, 1, 1, 2, 2, 3};
    Solution solution = Solution();
    cout << solution.removeDuplicates(input_v) << endl;
    for (auto num: input_v)
        cout << num << " ";
    cout << endl;
    return 0;
}
